### PR TITLE
Strict type check for all hooks in amazon

### PR DESCRIPTION
**Pull request by mlgruby**
Since the code base for Amazon is a little bit large I took an approach of going through one module at a time. 

@kaxil @mik-laj let me know if this approach is good or otherwise I will make this PR as WIP and include other modules as well. 

46.375552 amazon  ============> from 42 to 46 (This just includes hooks) 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/sns.py [hunk 1]
```diff
@@ -52,7 +53,7 @@ class AwsSnsHook(AwsBaseHook):
     def __init__(self, *args, **kwargs):
         super().__init__(client_type='sns', *args, **kwargs)
 
-    def publish_to_target(self, target_arn, message, subject=None, message_attributes=None):
+    def publish_to_target(self, target_arn, message, subject=None, message_attributes: Optional[dict] = None):
```
Comment (kaxil) [inline airflow/providers/amazon/aws/hooks/sns.py]: Can you add TypeHints to `target_arn`, `message` and `subject` too please

Comment (mlgruby) [reply airflow/providers/amazon/aws/hooks/sns.py]: Sure, I might have missed it. I will add those!

**Comment by mik-laj** [comment]
Should we use https://pypi.org/project/boto3-stubs/ or other similar packages?

CC: @coopergillan

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
> Should we use https://pypi.org/project/boto3-stubs/ or other similar packages?
> 
> CC: @coopergillan

Any comments here @coopergillan ?

**Review by mik-laj — APPROVED** [review]

**Comment by mik-laj** [comment]
@kaxil  We can add additional type hint later.

**Comment by kaxil** [comment]
Good work again here @mlgruby

**Comment by mik-laj** [comment]
I created issues about `boto3-stubs` package: https://github.com/apache/airflow/issues/11297

**Pull request merged**